src/utils/scipy_solver.py

```diff
@@ -0,0 +1,138 @@
+                   time_limit: float = float('inf'),
+                   ) -> dimod.SampleSet:
+        """Use HiGHS via SciPy to solve a constrained quadratic model.
+
+        Note that HiGHS requires the objective and constraints to be
+        linear.
+
+        Args:
+            cqm: A constrained quadratic model.
+            time_limit: The maximum time in seconds to search.
+
+        Returns:
+            A sample set with any solutions returned by
+            :func:`scipy.optimize.milp()`.
+
+        Raises:
+            ValueError: If the given constrained quadratic model contains
+                any quadratic terms.
+
+        """
+        # Note: we name the input variables according to SciPy's naming
+        # conventions
+
+        # Handle the empty case
+        if not cqm.variables:
+            return dimod.SampleSet.from_samples_cqm([], cqm, info=dict(run_time=0))
+
+        # Check that we're a linear model
+        if not cqm.objective.is_linear():
+            raise ValueError("scipy.optimize.milp() does not support objectives "
+                             "with quadratic interactions")
+        if not all(comp.lhs.is_linear() for comp in cqm.constraints.values()):
+            raise ValueError("scipy.optimize.milp() does not support constraints "
+                             "with quadratic interactions")
+
+        num_variables = cqm.num_variables()
+
+        # The objective
+        c = np.empty(num_variables, dtype=float)
+        for i, v in enumerate(cqm.variables):
+            c[i] = cqm.objective.linear.get(v, 0)
+
+        # The vartypes and the bounds
+        integrality = np.empty(num_variables, dtype=np.uint8)
+        lb = np.empty(num_variables, dtype=float)
+        ub = np.empty(num_variables, dtype=float)
+        for i, v in enumerate(cqm.variables):
+            vartype = cqm.vartype(v)
+            if vartype is dimod.BINARY:
+                integrality[i] = 1  # 1 indicates integer variable
+            elif vartype is dimod.INTEGER:
+                integrality[i] = 1  # 1 indicates integer variable
+            elif vartype is dimod.REAL:
+                integrality[i] = 0  # 0 indicates continuous
+            else:
+                raise ValueError("unexpected vartype")
+
+            lb[i] = cqm.lower_bound(v)
+            ub[i] = cqm.upper_bound(v)
+
+        # The constraints
+        constraints = list(SciPyCQMSolver.iter_constraints(cqm))
+
+        t = time.perf_counter()
+        solution = scipy.optimize.milp(c,
+                                       integrality=integrality,
+                                       bounds=scipy.optimize.Bounds(lb=lb, ub=ub),
+                                       options=dict(time_limit=time_limit),
+                                       constraints=constraints,
+                                       )
+        run_time = time.perf_counter() - t
+
+        # If we're infeasible, return an empty solution
+        if not solution.success:
```

Comment:
I did a bit more testing, and it looks like `not solution.success` when the solver times out, even if it has a valid solution. We should therefore do
```suggestion
        if solution.x is None:
```
which will handle the case that the solver timed out but it has an interim solution.